File: Data Structure/9. Graph/Adjacency Matrix/adjMatrix.py
# python implementation of DFS traversal of a graph
class Graph:
  # constructor
  def __init__(self, size):
    # use list comprehension to create empty adjacency list and 
    # visited boolean tracker
    self.num_nodes = size
    self.visited = [False for i in xrange(size)]
    self.adj_matrix = [[0 for j in xrange(size)] for i in xrange(size)]
  
  # add edges to the graph
  def add_edge(self, node1, node2):
    self.adj_matrix[node1][node2] = 1

  # dfs algorithm
  def dfs(self, start):
    print(str(start) + " "),
    self.visited[start] = True
    for j in range(self.num_nodes):
      if not self.visited[j] and self.adj_matrix[start][j]:
        self.dfs(j) # recursively visit connected node until return

# ================================
# testing the Graph class with DFS
# ================================

# create graph with 3 nodes
graph = Graph(4)

# add edges with nodes to the graph
graph.add_edge(0,1)
graph.add_edge(0,2)
graph.add_edge(1,0)
graph.add_edge(2,0)
graph.add_edge(2,3)
graph.add_edge(3,2)

# search the graph using dfs algorithm
print("Ajacency Matrix implementation of graph:")
graph.dfs(0)

File: Data Structures/9. Graph/Adjacency Matrix/adjMatrix.cpp
// a DFS implementation
# include <iostream>
# include <vector>
using namespace std;

// Graph class
class Graph {
  vector< vector<int> > adjacencyMatrix; // adjacency list implementation
  vector<bool> visited; // track visited node
  int nodeNum;
public:
  // constructor to resize the adjacencyMatrix and Visited vector
  Graph(int size) {
    nodeNum = size;
    adjacencyMatrix.resize(size);
    for (int i = 0; i < size; i++)
      adjacencyMatrix[i].resize(size);
    visited.resize(size);
  }

  // add adjacencyMatrix to the graph with given current node
  void addEdge(int node1, int node2) {
    adjacencyMatrix[node1][node2] = 1;
  }

  // Depth first search algorithm
  void DFS(int start) {
    cout << start << " ";
    visited[start] = true;
    for(int j = 0; j < nodeNum; j++)
        if(!visited[j] && adjacencyMatrix[start][j])
          DFS(j);
  }
};

int main() {
  cout << "Ajacency Matrix implementation of graph: \n";

  // create a graph called g and add adjacencyMatrix to it
  Graph graph(4);
  graph.addEdge(0,1);
  graph.addEdge(0,2);
  graph.addEdge(1,0);
  graph.addEdge(2,0);
  graph.addEdge(2,3);
  graph.addEdge(3,2);
  graph.DFS(0);
}